[v2.2.2] useAttrs type resolution failure

### Vue - Official extension or vue-tsc version

2.2.2

### VSCode version

1.97.2

### Vue version

3.5.13

### TypeScript version

5.7.3

### System Info

```shell
System:
    OS: Linux 5.15 Ubuntu 24.04.2 LTS 24.04.2 LTS (Noble Numbat)
    CPU: (2) x64 Intel(R) Core(TM) i5-4300U CPU @ 1.90GHz
    Memory: 919.50 MB / 4.81 GB
    Container: Yes
    Shell: 5.2.21 - /bin/bash
Binaries:
    Node: 22.14.0 - ~/.nvm/versions/node/v22.14.0/bin/node
    npm: 11.1.0 - ~/.nvm/versions/node/v22.14.0/bin/npm
    pnpm: 10.3.0 - ~/.local/share/pnpm/pnpm
Browsers:
    - (it is WSL)
```

### package.json dependencies

```json
{
  "dependencies": {
    "@nuxt/image": "^1.9.0",
    "nuxt": "^3.15.4",
    "vue": "^3.5.13",
    "vue-router": "^4.5.0"
  },
  "devDependencies": {
    "typescript": "^5.7.3",
    "vue-tsc": "^2.2.2"
  }
}
```

### Steps to reproduce

1. Navigate to the [reproduction](https://stackblitz.com/edit/github-g74ymcux-eyg4pnvv).
2. In terminal launch `pnpm typecheck`.
3. See the error.

### What is expected?

Typecheck to be passed without errors.

### What is actually happening?

Typecheck is failed with the error:
```bash
node_modules/.pnpm/@nuxt+image@1.9.0/node_modules/@nuxt/image/dist/runtime/components/NuxtImg.vue:41:7 -
 error TS7022: 'attrs' implicitly has type 'any' because it does not have a type annotation and is referenced directly or indirectly in its own initializer.

41 const attrs = useAttrs()
         ~~~~~


Found 1 error in node_modules/.pnpm/@nuxt+image@1.9.0/node_modules/@nuxt/image/dist/runtime/components/NuxtImg.vue:41


 ERROR  Process exited with non-zero status (2)                                      11:45:44 PM

    at new x (node_modules/.pnpm/tinyexec@0.3.2/node_modules/tinyexec/dist/main.js:440:5)
    at R._waitForOutput (node_modules/.pnpm/tinyexec@0.3.2/node_modules/tinyexec/dist/main.js:548:13)
    at async Object.run (node_modules/.pnpm/@nuxt+cli@3.21.1/node_modules/@nuxt/cli/dist/chunks/typecheck.mjs:58:7)
    at async runCommand (node_modules/.pnpm/citty@0.1.6/node_modules/citty/dist/index.mjs:337:16)
    at async runCommand (node_modules/.pnpm/citty@0.1.6/node_modules/citty/dist/index.mjs:328:11)
    at async Module.runMain (node_modules/.pnpm/citty@0.1.6/node_modules/citty/dist/index.mjs:466:7) 



 ERROR  Process exited with non-zero status (2)                                      11:45:44 PM

 ELIFECYCLE  Command failed with exit code 1.
```

### Link to minimal reproduction

https://stackblitz.com/edit/github-g74ymcux-eyg4pnvv

### Any additional comments?

I encountered this problem with the latest version of vue-tsc in the Nuxt project with `nuxt/image` module enabled. What is interesting is that typecheck reported an error for this line `const attrs = useAttrs()` in the `NuxtImg.vue` component only, while it is seem OK for the same line in the `App.vue` component (but I assume it is just the same as for `useSlots()`, when typecheck shows errors randomly, not for all occurrences).

--- a/packages/language-core/lib/codegen/template/index.ts
+++ b/packages/language-core/lib/codegen/template/index.ts
@@ -21,6 +21,8 @@ export interface TemplateCodegenOptions {
 	hasDefineSlots?: boolean;
 	slotsAssignName?: string;
 	propsAssignName?: string;
+	slotsReferenceNames: Set<string>;
+	attrsReferenceNames: Set<string>;
 	inheritAttrs: boolean;
 	selfComponentName?: string;
 }
@@ -33,6 +35,12 @@ export function* generateTemplate(options: TemplateCodegenOptions): Generator<Co
 	}
 	if (options.propsAssignName) {
 		ctx.addLocalVariable(options.propsAssignName);
+	}
+	for (const name of options.slotsReferenceNames) {
+		ctx.addLocalVariable(name);
+	}
+	for (const name of options.attrsReferenceNames) {
+		ctx.addLocalVariable(name);
 	}
 	const slotsPropertyName = getSlotsPropertyName(options.vueCompilerOptions.target);
 	ctx.specialVars.add(slotsPropertyName);

--- a/packages/language-core/lib/parsers/scriptSetupRanges.ts
+++ b/packages/language-core/lib/parsers/scriptSetupRanges.ts
@@ -54,11 +54,15 @@ type DefineOptions = {
 	inheritAttrs?: string;
 };
 
-type UseAttrs = CallExpressionRange;
+type UseAttrs = CallExpressionRange & {
+	name?: string;
+};
 
 type UseCssModule = CallExpressionRange;
 
-type UseSlots = CallExpressionRange;
+type UseSlots = CallExpressionRange & {
+	name?: string;
+};
 
 type UseTemplateRef = CallExpressionRange & {
 	name?: string;
@@ -284,26 +288,21 @@ export function parseScriptSetupRanges(
 			}
 			else if (vueCompilerOptions.macros.defineProps.includes(callText)) {
 				defineProps = {
-					...parseCallExpression(node),
+					...parseCallExpressionAssignment(node, parent),
 					statement: getStatementRange(ts, parents, node, ast),
 					argNode: node.arguments[0]
 				};
-				if (ts.isVariableDeclaration(parent)) {
-					if (ts.isObjectBindingPattern(parent.name)) {
-						defineProps.destructured = new Map();
-						const identifiers = collectIdentifiers(ts, parent.name, []);
-						for (const { id, isRest, initializer } of identifiers) {
-							const name = _getNodeText(id);
-							if (isRest) {
-								defineProps.destructuredRest = name;
-							}
-							else {
-								defineProps.destructured.set(name, initializer);
-							}
+				if (ts.isVariableDeclaration(parent) && ts.isObjectBindingPattern(parent.name)) {
+					defineProps.destructured = new Map();
+					const identifiers = collectIdentifiers(ts, parent.name, []);
+					for (const { id, isRest, initializer } of identifiers) {
+						const name = _getNodeText(id);
+						if (isRest) {
+							defineProps.destructuredRest = name;
 						}
-					}
-					else {
-						defineProps.name = _getNodeText(parent.name);
+						else {
+							defineProps.destructured.set(name, initializer);
+						}
 					}
 				}
 				else if (
@@ -327,12 +326,9 @@ export function parseScriptSetupRanges(
 			}
 			else if (vueCompilerOptions.macros.defineEmits.includes(callText)) {
 				defineEmits = {
-					...parseCallExpression(node),
+					...parseCallExpressionAssignment(node, parent),
 					statement: getStatementRange(ts, parents, node, ast)
 				};
-				if (ts.isVariableDeclaration(parent)) {
-					defineEmits.name = _getNodeText(parent.name);
-				}
 				if (node.typeArguments?.length && ts.isTypeLiteralNode(node.typeArguments[0])) {
 					for (const member of node.typeArguments[0].members) {
 						if (ts.isCallSignatureDeclaration(member)) {
@@ -347,12 +343,9 @@ export function parseScriptSetupRanges(
 			}
 			else if (vueCompilerOptions.macros.defineSlots.includes(callText)) {
 				defineSlots = {
-					...parseCallExpression(node),
+					...parseCallExpressionAssignment(node, parent),
 					statement: getStatementRange(ts, parents, node, ast)
 				};
-				if (ts.isVariableDeclaration(parent)) {
-					defineSlots.name = _getNodeText(parent.name);
-				}
 			}
 			else if (vueCompilerOptions.macros.defineExpose.includes(callText)) {
 				defineExpose = parseCallExpression(node);
@@ -377,22 +370,19 @@ export function parseScriptSetupRanges(
 				}
 			}
 			else if (vueCompilerOptions.composables.useAttrs.includes(callText)) {
-				useAttrs.push(parseCallExpression(node));
+				useAttrs.push(parseCallExpressionAssignment(node, parent));
 			}
 			else if (vueCompilerOptions.composables.useCssModule.includes(callText)) {
 				useCssModule.push(parseCallExpression(node));
 			}
 			else if (vueCompilerOptions.composables.useSlots.includes(callText)) {
-				useSlots.push(parseCallExpression(node));
+				useSlots.push(parseCallExpressionAssignment(node, parent));
 			}
 			else if (
 				vueCompilerOptions.composables.useTemplateRef.includes(callText)
 				&& !node.typeArguments?.length
 			) {
-				useTemplateRef.push({
-					name: ts.isVariableDeclaration(parent) ? _getNodeText(parent.name) : undefined,
-					...parseCallExpression(node)
-				});
+				useTemplateRef.push(parseCallExpressionAssignment(node, parent));
 			}
 		}
 
@@ -412,6 +402,13 @@ export function parseScriptSetupRanges(
 			exp: _getStartEnd(node.expression),
 			arg: node.arguments.length ? _getStartEnd(node.arguments[0]) : undefined,
 			typeArg: node.typeArguments?.length ? _getStartEnd(node.typeArguments[0]) : undefined,
+		};
+	}
+
+	function parseCallExpressionAssignment(node: ts.CallExpression, parent: ts.Node) {
+		return {
+			name: ts.isVariableDeclaration(parent) ? _getNodeText(parent.name) : undefined,
+			...parseCallExpression(node),
 		};
 	}
 

--- a/packages/language-core/lib/plugins/vue-tsx.ts
+++ b/packages/language-core/lib/plugins/vue-tsx.ts
@@ -69,19 +69,21 @@ export default plugin;
 
 function createTsx(
 	fileName: string,
-	_sfc: Sfc,
+	sfc: Sfc,
 	ctx: Parameters<VueLanguagePlugin>[0],
 	appendGlobalTypes: boolean
 ) {
 	const ts = ctx.modules.typescript;
+
 	const getLang = computed(() => {
-		return !_sfc.script && !_sfc.scriptSetup ? 'ts'
-			: _sfc.scriptSetup && _sfc.scriptSetup.lang !== 'js' ? _sfc.scriptSetup.lang
-				: _sfc.script && _sfc.script.lang !== 'js' ? _sfc.script.lang
+		return !sfc.script && !sfc.scriptSetup ? 'ts'
+			: sfc.scriptSetup && sfc.scriptSetup.lang !== 'js' ? sfc.scriptSetup.lang
+				: sfc.script && sfc.script.lang !== 'js' ? sfc.script.lang
 					: 'js';
 	});
+
 	const getResolvedOptions = computed(() => {
-		const options = parseVueCompilerOptions(_sfc.comments);
+		const options = parseVueCompilerOptions(sfc.comments);
 		if (options) {
 			const resolver = new CompilerOptionsResolver();
 			resolver.addConfig(options, path.dirname(fileName));
@@ -89,33 +91,37 @@ function createTsx(
 		}
 		return ctx.vueCompilerOptions;
 	});
+
 	const getScriptRanges = computed(() =>
-		_sfc.script
-			? parseScriptRanges(ts, _sfc.script.ast, !!_sfc.scriptSetup, false)
+		sfc.script
+			? parseScriptRanges(ts, sfc.script.ast, !!sfc.scriptSetup, false)
 			: undefined
 	);
+
 	const getScriptSetupRanges = computed(() =>
-		_sfc.scriptSetup
-			? parseScriptSetupRanges(ts, _sfc.scriptSetup.ast, getResolvedOptions())
+		sfc.scriptSetup
+			? parseScriptSetupRanges(ts, sfc.scriptSetup.ast, getResolvedOptions())
 			: undefined
 	);
+
 	const getSetupBindingNames = computedSet(
 		computed(() => {
 			const newNames = new Set<string>();
 			const bindings = getScriptSetupRanges()?.bindings;
-			if (_sfc.scriptSetup && bindings) {
+			if (sfc.scriptSetup && bindings) {
 				for (const { range } of bindings) {
-					newNames.add(_sfc.scriptSetup.content.slice(range.start, range.end));
+					newNames.add(sfc.scriptSetup.content.slice(range.start, range.end));
 				}
 			}
 			return newNames;
 		})
 	);
+
 	const getSetupImportComponentNames = computedSet(
 		computed(() => {
 			const newNames = new Set<string>();
 			const bindings = getScriptSetupRanges()?.bindings;
-			if (_sfc.scriptSetup && bindings) {
+			if (sfc.scriptSetup && bindings) {
 				for (const { range, moduleName, isDefaultImport, isNamespace } of bindings) {
 					if (
 						moduleName
@@ -123,13 +129,14 @@ function createTsx(
 						&& !isNamespace
 						&& ctx.vueCompilerOptions.extensions.some(ext => moduleName.endsWith(ext))
 					) {
-						newNames.add(_sfc.scriptSetup.content.slice(range.start, range.end));
+						newNames.add(sfc.scriptSetup.content.slice(range.start, range.end));
 					}
 				}
 			}
 			return newNames;
 		})
 	);
+
 	const getSetupDestructuredPropNames = computedSet(
 		computed(() => {
 			const newNames = new Set(getScriptSetupRanges()?.defineProps?.destructured?.keys());
@@ -140,6 +147,7 @@ function createTsx(
 			return newNames;
 		})
 	);
+
 	const getSetupTemplateRefNames = computedSet(
 		computed(() => {
 			const newNames = new Set(
@@ -150,29 +158,52 @@ function createTsx(
 			return newNames;
 		})
 	);
+
 	const setupHasDefineSlots = computed(() => !!getScriptSetupRanges()?.defineSlots);
+
 	const getSetupSlotsAssignName = computed(() => getScriptSetupRanges()?.defineSlots?.name);
+
 	const getSetupPropsAssignName = computed(() => getScriptSetupRanges()?.defineProps?.name);
+
 	const getSetupInheritAttrs = computed(() => {
 		const value = getScriptSetupRanges()?.defineOptions?.inheritAttrs ?? getScriptRanges()?.exportDefault?.inheritAttrsOption;
 		return value !== 'false';
 	});
+
+	const getSetupSlotsReferenceName = computedSet(
+		computed(() => {
+			const newNames = new Set(
+				getScriptSetupRanges()?.useSlots.map(({ name }) => name).filter(name => name !== undefined)
+			);
+			return newNames;
+		})
+	);
+
+	const getSetupAttrsReferenceName = computedSet(
+		computed(() => {
+			const newNames = new Set(
+				getScriptSetupRanges()?.useAttrs.map(({ name }) => name).filter(name => name !== undefined)
+			);
+			return newNames;
+		})
+	);
+
 	const getComponentSelfName = computed(() => {
 		const { exportDefault } = getScriptRanges() ?? {};
-		if (_sfc.script && exportDefault?.nameOption) {
+		if (sfc.script && exportDefault?.nameOption) {
 			const { nameOption } = exportDefault;
-			return _sfc.script.content.slice(nameOption.start + 1, nameOption.end - 1);
+			return sfc.script.content.slice(nameOption.start + 1, nameOption.end - 1);
 		}
 		const { defineOptions } = getScriptSetupRanges() ?? {};
-		if (_sfc.scriptSetup && defineOptions?.name) {
+		if (sfc.scriptSetup && defineOptions?.name) {
 			return defineOptions.name;
 		}
 		const baseName = path.basename(fileName);
 		return capitalize(camelize(baseName.slice(0, baseName.lastIndexOf('.'))));
 	});
-	const getGeneratedTemplate = computed(() => {
 
-		if (getResolvedOptions().skipTemplateCodegen || !_sfc.template) {
+	const getGeneratedTemplate = computed(() => {
+		if (getResolvedOptions().skipTemplateCodegen || !sfc.template) {
 			return;
 		}
 
@@ -181,7 +212,7 @@ function createTsx(
 			ts,
 			compilerOptions: ctx.compilerOptions,
 			vueCompilerOptions: getResolvedOptions(),
-			template: _sfc.template,
+			template: sfc.template,
 			edited: getResolvedOptions().__test || (fileEditTimes.get(fileName) ?? 0) >= 2,
 			scriptSetupBindingNames: getSetupBindingNames(),
 			scriptSetupImportComponentNames: getSetupImportComponentNames(),
@@ -190,12 +221,13 @@ function createTsx(
 			hasDefineSlots: setupHasDefineSlots(),
 			slotsAssignName: getSetupSlotsAssignName(),
 			propsAssignName: getSetupPropsAssignName(),
+			slotsReferenceNames: getSetupSlotsReferenceName(),
+			attrsReferenceNames: getSetupAttrsReferenceName(),
 			inheritAttrs: getSetupInheritAttrs(),
 			selfComponentName: getComponentSelfName(),
 		});
 
 		let current = codegen.next();
-
 		while (!current.done) {
 			const code = current.value;
 			codes.push(code);
@@ -207,15 +239,17 @@ function createTsx(
 			codes,
 		};
 	});
+
 	const getGeneratedScript = computed(() => {
-		const codes: Code[] = [];
 		const linkedCodeMappings: Mapping[] = [];
 		let generatedLength = 0;
+
+		const codes: Code[] = [];
 		const codegen = generateScript({
 			ts,
 			compilerOptions: ctx.compilerOptions,
 			vueCompilerOptions: getResolvedOptions(),
-			sfc: _sfc,
+			sfc: sfc,
 			edited: getResolvedOptions().__test || (fileEditTimes.get(fileName) ?? 0) >= 2,
 			fileName,
 			lang: getLang(),
@@ -231,7 +265,6 @@ function createTsx(
 		fileEditTimes.set(fileName, (fileEditTimes.get(fileName) ?? 0) + 1);
 
 		let current = codegen.next();
-
 		while (!current.done) {
 			const code = current.value;
 			codes.push(code);

